### PR TITLE
feat: implement KongLicense controller and inject to translator

### PR DESCRIPTION
**Pull request by randmonkey**
**What this PR does / why we need it**:

The PR implements a `KongLicense` controller which picks the newest `KongLicense` in the cluster and provides a `GetLicense` method for translator to inject Kong license into Kong configurations. 
The controller is enabled when `--enable-controller-kong-license` is set to `true` (by default) and sync license with Konnect disabled. When sync license with Konnect is enabled, the licenses from Konnect takes the precedence.

**Which issue this PR fixes**:

Major part of #5437.
**Special notes for your reviewer**:

Not fully finished yet, but can reconcile `KongLicense` resources and ready to receive some reviews. Major remaining part:
- When no `KongLicense` CRD in cluster, the controller will exit abnormally
- Maybe add validation of `KongLicense`?
- Disable applying to Kong gateway when it is OSS (in another issue/PR?)

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/controllers/configuration/konglicense_controller.go [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+	ConditionReasonReplacedByNewer = "ReplacedByNewer"
+	// maxConditionNum is the maximum number of condition items in the controller status.
+	maxConditionNum = 8
+)
+
+var _ controllers.Reconciler = &KongV1Alpha1KongLicenseReconciler{}
+
+func NewLicenseCache() cache.Store {
+	return cache.NewStore(kongLicenseKeyFunc)
+}
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/controllers/configuration/konglicense_controller.go:74]: Because the licenses are not processed inside the translator but used as a license injector, so I do not want to add them into the general "big cache".

File: internal/manager/run.go [hunk 2]
```diff
@@ -313,6 +314,24 @@ func Run(
...
+		setupLog.Info("Starting KongLicense controller")
+		licenseController := &configuration.KongV1Alpha1KongLicenseReconciler{
+			Client:           mgr.GetClient(),
+			Log:              ctrl.LoggerFrom(ctx).WithName("controllers").WithName("KongLicense"),
+			Scheme:           mgr.GetScheme(),
+			LicenseCache:     configuration.NewLicenseCache(),
+			CacheSyncTimeout: c.CacheSyncTimeout,
+			StatusQueue:      kubernetesStatusQueue,
+			ControllerName:   c.LeaderElectionID,
+		}
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/manager/run.go:328]: Will do.

File: internal/controllers/configuration/konglicense_controller.go [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+			// Then pick the effective license in KongLicenses remaining in cache.
+			chosenLicense := r.pickLicenseInCache()
+			r.setChosenLicense(chosenLicense)
+			if chosenLicense != nil {
+				log.V(util.DebugLevel).Info("Picked KongLicense remaining in cache", "name", chosenLicense.Name)
+				err := r.ensureControllerStatusProgrammedCondition(ctx, chosenLicense, metav1.ConditionTrue, ConditionReasonPickedAsLatest, "")
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+			}
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/controllers/configuration/konglicense_controller.go:187]: Sounds good, will do.

File: internal/controllers/configuration/konglicense_controller.go [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+	programmedStatus metav1.ConditionStatus,
+	reason string, message string,
+) error {
+	// Get the latest status of target KongLicense.
+	err := r.Client.Get(ctx, k8stypes.NamespacedName{Name: license.Name}, license)
+	if err != nil {
+		return fmt.Errorf("failed to get latest version of KongLicense %s: %w", license.Name, err)
+	}
+
+	fullControllerName := LicenseControllerType + ":" + r.ControllerName
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/controllers/configuration/konglicense_controller.go:284]: I used a different separator because the `LicenseControllerType` already contains a `/`. For constructing a valid URL, will accept the suggestion.

File: internal/controllers/configuration/konglicense_controller.go [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+	l, ok := obj.(*kongv1alpha1.KongLicense)
+	if !ok {
+		return "", fmt.Errorf("object is type %T, not KongLicense", obj)
+	}
+	return l.Name, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KongV1Alpha1KongLicenseReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// TODO: it fails when KongLicense CRD is absent. How should we fix it? By wrapping with DynamicCRDController?
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/controllers/configuration/konglicense_controller.go:86]: OK, will move to `controllerDef`s and this seems to resolve the problem.

File: internal/controllers/configuration/konglicense_controller.go [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+
+// GetLicense is the interface to get the license in Kong configuration format to use in translator.
+func (r *KongV1Alpha1KongLicenseReconciler) GetLicense() mo.Option[kong.License] {
+	chosenLicense := r.getChosenLicense()
+	if chosenLicense == nil {
+		r.Log.V(util.DebugLevel).Info("No KongLicense available")
+		return mo.None[kong.License]()
+	}
+	r.Log.V(util.DebugLevel).Info("Get license from KongLicense resource", "name", chosenLicense.Name)
+	// TODO: Validate KongLicense on Kong gateway.
```
Comment (czeslavo) [inline internal/controllers/configuration/konglicense_controller.go:343]: This should rather happen in the webhook, no?

File: internal/controllers/configuration/konglicense_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (czeslavo) [reply internal/controllers/configuration/konglicense_controller.go:74]: Ah, I see, the reconciler implements `GetLicense` and it's going to be used in the `Translator`. Let's make sure that we draw the boundary between the reconciler and the translator well by defining an interface. 👍

File: internal/controllers/configuration/konglicense_controller.go [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/util/kubernetes/object/status"
+	kongv1alpha1 "github.com/kong/kubernetes-ingress-controller/v3/pkg/apis/configuration/v1alpha1"
+)
+
+// -----------------------------------------------------------------------------
+// KongV1Alpha1 KongLicense - Reconciler
+// -----------------------------------------------------------------------------
+
+// KongV1Alpha1KongLicenseReconciler reconciles KongLicense resources.
+type KongV1Alpha1KongLicenseReconciler struct {
```
Comment (czeslavo) [inline internal/controllers/configuration/konglicense_controller.go:40]: Please make sure we have unit coverage for the reconciler (and/or its helpers). It could be an envtest to verify the reconciler behavior testing different scenarios aiming to cover all possible edge cases (two or more licenses, adding a newer one, etc.). 

The integration test could be used just to ensure the happy path - i.e. adding a valid `KongLicense` and ensuring it's been propagated to Kong.

**Comment by czeslavo** [comment]
BTW it might make sense to not include integration tests and translator code in this PR to not make it bigger as it's already 2k+ lines. :)

File: internal/controllers/configuration/konglicense_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/controllers/configuration/konglicense_controller.go]: Done. Moved to a `repickLicenseOnDelete` method.

File: internal/controllers/configuration/konglicense_controller.go (hunk 6 above)
Comment (randmonkey) [reply internal/controllers/configuration/konglicense_controller.go]: Yes, I will implement in a webhook later (may not be in 3.1?)

File: internal/controllers/configuration/konglicense_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/controllers/configuration/konglicense_controller.go]: Done.

File: internal/controllers/configuration/konglicense_controller.go (hunk 7 above)
Comment (randmonkey) [reply internal/controllers/configuration/konglicense_controller.go:40]: OK, will add tests in a new PR.

File: internal/manager/run.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (czeslavo) [reply internal/manager/run.go:328]: In my opinion, it's better to pass this parameter down even for the single controller but keep the consistency of initializing all controllers in `setupControllers`.

File: internal/controllers/configuration/konglicense_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (czeslavo) [reply internal/controllers/configuration/konglicense_controller.go:86]:
```suggestion
```

File: internal/controllers/configuration/konglicense_controller.go [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+
+// GetLicense is the interface to get the license in Kong configuration format to use in translator.
+func (r *KongV1Alpha1KongLicenseReconciler) GetLicense() mo.Option[kong.License] {
+	chosenLicense := r.getChosenLicense()
+	if chosenLicense == nil {
+		r.Log.V(util.DebugLevel).Info("No KongLicense available")
+		return mo.None[kong.License]()
+	}
+	r.Log.V(util.DebugLevel).Info("Get license from KongLicense resource", "name", chosenLicense.Name)
+	// TODO: Validate KongLicense on Kong gateway.
```
Comment (czeslavo) [inline internal/controllers/configuration/konglicense_controller.go:217]: Please consider either removing the TODO comment or supplementing it with a tracking issue link.

File: internal/controllers/configuration/konglicense_controller.go [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+		license.Status.KongLicenseControllerStatuses[controllerIndex].Conditions[conditionIndex] = wantedCondition
+		return r.Client.Status().Update(ctx, license)
+	}
+
+	return nil
+}
+
+// WrapKongLicenseReconcilerToDynamicCRDController wraps KongLicenseReconciler to DynamicCRDController
+// to watch precense of KongLicense CRD to avoid aborts if KongLicense is not installed when controller initialized.
+// REVIEW: Is there a better way to resolve that?
```
Comment (czeslavo) [inline internal/controllers/configuration/konglicense_controller.go:364]: So far we've been using `DynamicCRDController` only to allow Gateway API CRDs to be installed during runtime (the only non-GW API CRD is `KongUpstreamPolicyReconciler` although it depends on `HTTPRoute` so it's also wrapped in that). 

For regular independent Kong CRDs we were rather expecting the CRD to be installed and cache sync to fail if it's not. As `KongLicense` is introduced as `alpha`-maturity feature, I think it could be irritating to users if they upgrade (forgetting to upgrade CRDs) and get weird crashes because of cache sync timeouts so I'm ok with going with the `DynamicCRDController` for that.

Another option would be to make this feature-gated and enable the controller only if the feature gate is enabled, making it non-breaking for people not using the `KongLicense` straight away.

File: internal/manager/run.go [hunk 10]
```diff
@@ -313,6 +314,27 @@ func Run(
...
+		dynamicLicenseController := configuration.WrapKongLicenseReconcilerToDynamicCRDController(
+			ctx, mgr, licenseController,
+		)
+		err := dynamicLicenseController.SetupWithManager(mgr)
+		if err != nil {
+			return fmt.Errorf("failed to start KongLicense controller: %w", err)
+		}
+		configTranslator.InjectLicenseGetter(licenseController)
```
[the remark this answers is not in the document]
Comment (czeslavo) [reply internal/manager/run.go:336]: Ah, I see - we do not enable the `KongLicense` controller when the Konnect license agent is enabled. I think it would be worth having this covered in some unit test so it's not messed up in future refactors. We could extract a helper setup function (e.g. `setupLicenseGetter`) that we could test in isolation and verify the logic responsible for building the controller or the konnect agent.

File: internal/controllers/configuration/konglicense_controller.go (hunk 8 above)
Comment (randmonkey) [reply internal/controllers/configuration/konglicense_controller.go]: Done. Added issue #5566.

File: internal/manager/run.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/manager/run.go]: Moved to a `setupLicenseGetter` function in `setup.go`.

**Review by czeslavo — APPROVED** [review]
